File: CPI.R
#' Current Precipitation Index (CPI)
#'
#' As described in:
#'  Smakhtin, V. Y., Masse, B. 2000. Continuous 
#'  daily hydrograph simulation using duration
#'  curves of a precipitation index.Hydrological
#'  Processes 14: 1083-1100.
#'  
#'  Original MATLAB: Reggie Walters, SFPUC, 2016
#'  Translated R: Don Park, UMass Team, 2020
#'  
#'  @param prcp Instantaneous precipitation time series vector
#'  @param beta Recession coefficient
#'  @param time Time array corresponding to precipitation input series 
#'  
#'  @return Dataframe with CPI and Time
#'  @export

CPI <- function(prcp, beta, time) {
  
  c_i <- rep(0, length(prcp))
  
  for(i in 2:length(c_i)) {
    c_i[i] <- prcp[i-1] + beta * c_i[i-1]
  }
  
  return(data.frame(time = time, ci = c_i))
  
  
}

File: CPI.m
function [c_i, dt] = CPI(Precip, beta, time)
% current precipitation index (CPI) calculation routine
% as desribed in:
% Smakhtin, V. Y., Masse, B. 2000. Continuous daily hydrograph simulation
% using duration curves of a precipitation index. Hydrological Processes
% 14: 1083-1100.
% r. walters, sfpuc, june 2020
%
% INPUT:
%       Precip: instantaneous precipitation time series vector
%       beta:   recession coefficient
%       time:   time array corresponding to precipitation input series
%

% initialize container (all zeroes) for function output variable, ci
c_i = 0 .* ones(length(Precip),1);          

for t = 2:length(c_i)
    c_i(t) = Precip(t) + beta * c_i(t-1);
end

dt = time;
